### PR TITLE
refactor CapabilitiesResolver to support reference mode storage keys

### PR DESCRIPTION
**Pull request by mmandlis**
mostly replication of https://github.com/PolymerLabs/arcs/pull/4773 in TS

### CODE REVIEW

File: src/runtime/capabilities-resolver.ts [hunk 1]
```diff
@@ -13,11 +13,43 @@ import {ArcId} from './id.js';
...
 
-export type StorageKeyOptions = Readonly<{
+export type CapabilitiesResolverOptions = Readonly<{
   arcId: ArcId;
 }>;
 
+export abstract class StorageKeyOptions {
+  constructor(public readonly arcId: ArcId,
+              public readonly schemaHash: string,
+              protected readonly schemaName: string = null) {}
```
Comment (csilvestrini) [inline src/runtime/capabilities-resolver.ts]: Nit: as far as I am aware, we don't usually indent like this in our TypeScript code.

```
  constructor(
      public readonly arcId: ArcId,
      public readonly schemaHash: string,
      protected readonly schemaName: string = null) {}
```

File: src/runtime/capabilities-resolver.ts [hunk 2]
```diff
@@ -13,11 +13,43 @@ import {ArcId} from './id.js';
...
+  abstract location():string;
```
Comment (csilvestrini) [inline src/runtime/capabilities-resolver.ts]: Nit: insert space after colon

File: src/runtime/tests/capabilities-resolver-test.ts [hunk 3]
```diff
@@ -20,50 +23,83 @@ import {Runtime} from '../runtime.js';
...
+    assert.instanceOf(refKey.storageKey, expectedType);
+  }
+  async function verifyThrows(promise: Promise<StorageKey>) {
+    try {
+      await promise;
+      assert.fail();
+    } catch (e) {
+      // do nothing.
+    }
+  }
```
Comment (csilvestrini) [inline src/runtime/tests/capabilities-resolver-test.ts]: You can just use `assertThrowsAsync` from `src/testing/test-util.ts` instead.

File: src/runtime/capabilities-resolver.ts (hunk 1 above)
Comment (mmandlis) [reply src/runtime/capabilities-resolver.ts]: done

File: src/runtime/capabilities-resolver.ts (hunk 2 above)
Comment (mmandlis) [reply src/runtime/capabilities-resolver.ts]: done

File: src/runtime/tests/capabilities-resolver-test.ts (hunk 3 above)
Comment (mmandlis) [reply src/runtime/tests/capabilities-resolver-test.ts]: oh, thank you! i knew there should be something like that, but couldn't find it...

**Pull request merged**